### PR TITLE
More type erasure for AnySegmentIterable to reduce compile times in Debug build

### LINKED ISSUE #1145: for_each function on segment
```c++
resolve_segment_type<T>(*segment, [&](auto& typed_segment) {
  auto iterable = create_iterable_from_segment<T>(typed_segment);
  iterable.for_each([&](const auto& value) {
```

@mrks:
> The three lines above look like boilerplate. Thinking about it, why do I first have to create an iterable
> Shouldn't it be possible to call for_each on the segment?
> (If this makes sense, it should be its own issue)

### PR DESCRIPTION
**Pull request by mrzzzrm**
Fix #1145

This introduces `segment_iterate()` and `segment_with_iterators()` as your new way to go to iterate over a segment's data. They compile way faster in debug builds than anything we had previously, as the passed functor is only instantiated once per `DataType`. In release builds, everything remains roughly the same.

I changed the JoinNestedLoop, which took ~15 Minutes to compile in release` to retrieve the outer value via a virtual function call, which greatly reduces its compile time.

ClangDebug libhyrise objects (accumulated):

```
ObjectCount 258 -> 260
BuildTime 632.07s -> 545.41s (-86.66s -13.71%)
FileSize 937.2MiB -> 645.2MiB (-292.0MiB -31.16%)
```

ClangRelease libhyrise objects (accumulated):

```
ObjectCount 258 -> 260
BuildTime 2505.70s -> 1579.45s (-926.25s -36.97%)
FileSize 106.7MiB -> 88.2MiB (-18.5MiB -17.35%)
```

### CODE REVIEW

**Comment by mrzzzrm** [comment]
#1337 :muscle:

File: src/lib/operators/jit_operator/operators/jit_read_tuples.cpp [hunk 1]
```diff
@@ -73,18 +73,18 @@ void JitReadTuples::before_chunk(const Table& in_table, const Chunk& in_chunk, J
...
-          context.inputs.push_back(
-              std::make_shared<JitSegmentReader<IteratorType, Type, false>>(it, input_column.tuple_value));
-        }
-      });
+
+    segment_with_iterators(*segment, [&](auto it, const auto end) {
+      using Type = typename decltype(it)::ValueType;
+      using IteratorType = decltype(it);
+
+      if (is_nullable) {
```
Comment (mrks) [inline src/lib/operators/jit_operator/operators/jit_read_tuples.cpp]: Not part of this PR, but I wonder how well the branch prediction works here. As this will not change for a long time, can the CPU execute this as efficiently as if the `if` was outside the loop?

File: src/lib/operators/join_mpsm/column_materializer_numa.hpp [hunk 2]
```diff
@@ -123,33 +124,32 @@ class ColumnMaterializerNUMA {
 
     return std::make_shared<JobTask>(
         [this, &output, &null_rows_output, segment, chunk_id, alloc, numa_node_id] {
-          resolve_segment_type<T>(*segment, [&](auto& typed_segment) {
-            _materialize_segment(typed_segment, chunk_id, null_rows_output, (*output)[numa_node_id]);
-          });
+          if (const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<T>>(segment)) {
+            _materialize_dictionary_segment(*dictionary_segment, chunk_id, null_rows_output, (*output)[numa_node_id]);
```
Comment (mrks) [inline src/lib/operators/join_mpsm/column_materializer_numa.hpp]: What is the advantage of this change? To me it looks like we now need more code and a second method (instead of an overloaded one)

File: src/lib/operators/join_nested_loop.cpp [hunk 3]
```diff
@@ -91,39 +92,38 @@ void JoinNestedLoop::_join_two_untyped_segments(const std::shared_ptr<const Base
...
-
-        iterable_left.with_iterators(
-            [&params_copy, &iterable_right, chunk_id_left_copy, chunk_id_right_copy](auto left_it, auto left_end) {
-              iterable_right.with_iterators([&](auto right_it, auto right_end) {
-                with_comparator(params_copy.predicate_condition, [&](auto comparator) {
-                  join_two_typed_segments(comparator, left_it, left_end, right_it, right_end, chunk_id_left_copy,
-                                          chunk_id_right_copy, params_copy);
-                });
-              });
+  segment_with_iterators<ResolveDataTypeTag, SegmentIterationTypeErasure::Always>(
```
Comment (mrks) [inline src/lib/operators/join_nested_loop.cpp]: This code is also used as a fallback by JoinIndex. As that operator is not used by default right now, this is not an issue, but we should probably add a comment over there saying that even for the nested loop join, this is not the most efficient way to do things.

File: src/lib/operators/join_nested_loop.cpp [hunk 4]
```diff
@@ -91,39 +92,38 @@ void JoinNestedLoop::_join_two_untyped_segments(const std::shared_ptr<const Base
...
+
+          // make sure that we do not compile invalid versions of these lambdas
+          constexpr auto LEFT_IS_STRING_COLUMN = (std::is_same<LeftType, std::string>{});
+          constexpr auto RIGHT_IS_STRING_COLUMN = (std::is_same<RightType, std::string>{});
+
+          constexpr auto NEITHER_IS_STRING_COLUMN = !LEFT_IS_STRING_COLUMN && !RIGHT_IS_STRING_COLUMN;
+          constexpr auto BOTH_ARE_STRING_COLUMN = LEFT_IS_STRING_COLUMN && RIGHT_IS_STRING_COLUMN;
+
+          if constexpr (NEITHER_IS_STRING_COLUMN || BOTH_ARE_STRING_COLUMN) {
+            // Dirty hack to avoid https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86740
```
Comment (mrks) [inline src/lib/operators/join_nested_loop.cpp]: This could be so nice and easy if it wasn't for gcc. :(

File: src/lib/operators/sort.cpp [hunk 5]
```diff
@@ -215,16 +216,12 @@ class Sort::SortImpl : public AbstractReadOnlyOperatorImpl {
...
-            null_value_rows.emplace_back(RowID{chunk_id, value.chunk_offset()}, SortColumnType{});
-          } else {
-            row_id_value_vector.emplace_back(RowID{chunk_id, value.chunk_offset()}, value.value());
-          }
-        });
+      segment_for_each<SortColumnType>(*base_segment, [&](const auto& value) {
+        if (value.is_null()) {
+          null_value_rows.emplace_back(RowID{chunk_id, value.chunk_offset()}, SortColumnType{});
+        } else {
+          row_id_value_vector.emplace_back(RowID{chunk_id, value.chunk_offset()}, value.value());
```
Comment (mrks) [inline src/lib/operators/sort.cpp]: I don't really like the `value.value()` part. We use this in other places as well. In the scan, I have called it `iterator_value`, which I am not very happy with either. Provided that we can come up with a better name for this, we should open an issue to make it consistent.

File: src/lib/operators/table_scan/column_between_table_scan_impl.hpp [hunk 6]
```diff
@@ -32,10 +32,12 @@ class ColumnBetweenTableScanImpl : public AbstractSingleColumnTableScanImpl {
...
 
-  void _scan_segment(const BaseSegment& segment, const ChunkID chunk_id, PosList& matches,
-                     const std::shared_ptr<const PosList>& position_filter) const;
-  void _scan_segment(const BaseDictionarySegment& segment, const ChunkID chunk_id, PosList& matches,
-                     const std::shared_ptr<const PosList>& position_filter) const;
+  void _scan_non_dictionary_segment(const BaseSegment& segment, const ChunkID chunk_id, PosList& matches,
+                                    const std::shared_ptr<const PosList>& position_filter) const;
+
+  // Optimized scan on DictionarySegments
+  void _scan_dictionary_segment(const BaseDictionarySegment& segment, const ChunkID chunk_id, PosList& matches,
```
Comment (mrks) [inline src/lib/operators/table_scan/column_between_table_scan_impl.hpp]: Argh, I was happy to find a way where we do not have to add types to the method name and where the fallback was implicit, not explicit. What is the advantage of this?

File: src/lib/operators/table_scan/column_is_null_table_scan_impl.cpp [hunk 7]
```diff
@@ -29,43 +30,29 @@ std::shared_ptr<PosList> ColumnIsNullTableScanImpl::scan_chunk(const ChunkID chu
...
-void ColumnIsNullTableScanImpl::_scan_segment(const BaseSegment& segment, const ChunkID chunk_id, PosList& matches,
-                                              const std::shared_ptr<const PosList>& position_filter) const {
-  resolve_data_and_segment_type(segment, [&](const auto type, const auto& typed_segment) {
-    using Type = typename decltype(type)::type;
-    auto iterable = create_iterable_from_segment<Type>(typed_segment);
-
+void ColumnIsNullTableScanImpl::_scan_non_value_segment(const BaseSegment& segment, const ChunkID chunk_id,
+                                                        PosList& matches,
+                                                        const std::shared_ptr<const PosList>& position_filter) const {
+  segment_with_iterators(segment, position_filter, [&](auto it, const auto end) {
```
Comment (mrks) [inline src/lib/operators/table_scan/column_is_null_table_scan_impl.cpp]: I had to read this a couple of times before I saw that `segment_with_iterators` also takes a position filter. Maybe we should include that in the name?

File: src/lib/CMakeLists.txt [hunk 8]
```diff
@@ -458,15 +458,18 @@ set(
     storage/run_length_segment.hpp
```
Comment (mrzzzrm) [inline src/lib/CMakeLists.txt]: (Random line)

Worst offenders in clang release build:

```
column_vs_column_table_scan_impl.cpp.o: 23.5MiB 476.156s
join_nested_loop.cpp.o: 3.6MiB 140.837s
expression_evaluator.cpp.o: 11.8MiB 105.403s
aggregate.cpp.o: 6.0MiB 57.92s
join_hash.cpp.o: 4.1MiB 39.615s
```

Comment (mrzzzrm) [reply src/lib/CMakeLists.txt]: Worst offenders in clang debug build
```
expression_evaluator.cpp.o: 54.2MiB 19.927s
server_session_test.cpp.o: 36.3MiB 11.934s
segment_statistics.cpp.o: 21.0MiB 9.539s
aggregate.cpp.o: 32.0MiB 9.272s
join_hash_steps_test.cpp.o: 28.4MiB 8.923s
any_segment_iterable.cpp.o: 23.0MiB 8.207s
```

File: src/lib/expression/evaluation/expression_evaluator.cpp [hunk 9]
```diff
@@ -1480,19 +1497,31 @@ std::vector<std::shared_ptr<ExpressionResult<Result>>> ExpressionEvaluator::_pru
...
+    auto chunk_offset = ChunkOffset{0};
 
     if (table->column_is_nullable(ColumnID{0})) {
-      result_nulls.reserve(table->row_count());
+      result_nulls.resize(table->row_count());
 
       for (auto chunk_id = ChunkID{0}; chunk_id < table->chunk_count(); ++chunk_id) {
         const auto& result_segment = *table->get_chunk(chunk_id)->get_segment(ColumnID{0});
-        materialize_nulls<Result>(result_segment, result_nulls);
+        segment_for_each<Result>(result_segment, [&](const auto& value) {
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/expression/evaluation/expression_evaluator.cpp]: Renamed

File: src/lib/operators/jit_operator/operators/jit_read_tuples.cpp (hunk 1 above)
Comment (mrzzzrm) [reply src/lib/operators/jit_operator/operators/jit_read_tuples.cpp]: No expertise in branch prediction, but I'd expect it to have learned that "is_nullable" is either true or false after a few iterations? Should I extract the condition?

File: src/lib/operators/join_nested_loop.cpp (hunk 3 above)
Comment (mrzzzrm) [reply src/lib/operators/join_nested_loop.cpp]:
```
  /**
   * The nested loops.
   *
   * The value in the outer loop is retrieved via virtual function calls ("SegmentIterationTypeErasure::Always") and only
   * the inner loop gets inlined. This is to keep the compile time of the JoinNestedLoop *somewhat* at bay, if we inline
   * both the inner and the outer loop, the JoinNestedLoop becomes the most expensive-to-compile file in all of Hyrise
   * by a margin
   */
```

Glad to discuss the JNL in person. As it stands it's a major pain compile time wise and - as you say - we don't even use it right now. We could consider other solutions, including materializing both columns before joining. This would actually avoid repeatedly de-coding the inner relation....

File: src/lib/operators/sort.cpp (hunk 5 above)
Comment (mrzzzrm) [reply src/lib/operators/sort.cpp]: I do not have a major problem with `value.value()`. Whatever you want this to be named like, it shall be named.

File: src/lib/operators/table_scan/column_is_null_table_scan_impl.cpp (hunk 7 above)
Comment (mrzzzrm) [reply src/lib/operators/table_scan/column_is_null_table_scan_impl.cpp]: Even though this made me spell out the hideous word "position_filter", I renamed it to "segment_with_iterators_and_position_filter" ;)

File: src/lib/operators/join_mpsm/column_materializer_numa.hpp (hunk 2 above)
Comment (mrzzzrm) [reply src/lib/operators/join_mpsm/column_materializer_numa.hpp]: Before, this code instantiated the lambda unnecessarily often.

This particular case is funny, because the let's-call-both-methods-_materialize_dictionary_segment-approach combined with a signature mismatch (probably due to a c/p mistake) led to the optimized dictionary variant never being called.

File: src/lib/operators/table_scan/column_is_null_table_scan_impl.cpp (hunk 7 above)
Comment (mrks) [reply src/lib/operators/table_scan/column_is_null_table_scan_impl.cpp]: I hate names. Maybe we should compile everything, decompile it, and live with random names.

`segment_with_iterators_filtered`?

Comment (mrzzzrm) [reply src/lib/operators/table_scan/column_is_null_table_scan_impl.cpp]: Maybe that's a tiny bit better...

File: src/lib/operators/table_scan/column_between_table_scan_impl.hpp (hunk 6 above)
Comment (mrks) [reply src/lib/operators/table_scan/column_between_table_scan_impl.hpp]: > First, the previous version instantiated the lambda for each segment type, only to distinguish between dictionary and non-dictionary segments. Minor, but this is why I first changed it.

Did this show up in the compile time and if so, how much was the difference?

At least you convinced me to think about this a bit more. Generally, I am not a fan of the `if (dynamic_cast...` pattern

Comment (mrks) [reply src/lib/operators/table_scan/column_between_table_scan_impl.hpp]: Should we decide that this is the way to go, I would whine about the use of raw pointers next. Is there any need to convert the shared_ptr to a raw pointer to a reference?

File: src/lib/operators/join_nested_loop.cpp (hunk 3 above)
Comment (mrks) [reply src/lib/operators/join_nested_loop.cpp]: For the index join, we should find a better solution anyway. Using the nested loop join here is just a quick fix.

**Comment by mrzzzrm** [comment]
Apparently there is a performance regression in the JoinHash, so WIP this again.

File: src/lib/operators/table_scan/column_between_table_scan_impl.hpp (hunk 6 above)
Comment (mrzzzrm) [reply src/lib/operators/table_scan/column_between_table_scan_impl.hpp]: > Did this show up in the compile time and if so, how much was the difference?

Nah, this was just me purging "resolve_*" stuff from the code base, I don't think this particular case would have had a negative impact, its just ~7 instantiations of a lambda

> Generally, I am not a fan of the if (dynamic_cast... pattern

Neither am I, but I think in this case this is better than using resolve_segment_type, just to distinguish between two types of segments


> Should we decide that this is the way to go, I would whine about the use of raw pointers next. Is there any need to convert the shared_ptr to a raw pointer to a reference?

This doesn't really bother me. We could hand the shared_ptr all the way through the TableScanImpl, but I reckon that's more typing in the end.

Comment (mrzzzrm) [reply src/lib/operators/table_scan/column_between_table_scan_impl.hpp]: Probably misunderstand you, but "dynamic_pointer_cast" won't work on references, right? So if we stick with passing "BaseSegment" around as reference (which I am in favour of), then we have no choice but to use "dynamic_cast<x*>", since "dynamic_cast<x&>" reports failures with an exception.

**Comment by mrzzzrm** [comment]
Performance regression fixed, `it++` is dangerous.

```
+-----------+-----------------+------+-----------------+------+--------+---------------------------------+
| Benchmark | prev. iter/s    | runs | new iter/s      | runs | change | p-value (significant if <0.001) |
+-----------+-----------------+------+-----------------+------+--------+---------------------------------+
| TPC-H 1   | 2.32604908943   | 140  | 2.35022330284   | 142  | +1%    |     (run time too short) 0.0211 |
| TPC-H 2   | 38.4405822754   | 2307 | 38.5891456604   | 2316 | +0%    |     (run time too short) 0.0000 |
| TPC-H 3   | 20.9787883759   | 1259 | 20.7794380188   | 1247 | -1%    |     (run time too short) 0.0000 |
| TPC-H 4   | 8.18517112732   | 492  | 8.67770385742   | 521  | +6%    |     (run time too short) 0.0000 |
| TPC-H 5   | 25.6755504608   | 1541 | 25.6734676361   | 1541 | -0%    |     (run time too short) 0.9229 |
| TPC-H 6   | 53.4845466614   | 3210 | 53.6534576416   | 3220 | +0%    |     (run time too short) 0.0028 |
| TPC-H 7   | 8.20349502563   | 493  | 8.24221611023   | 495  | +0%    |     (run time too short) 0.9104 |
| TPC-H 8   | 47.3035392761   | 2839 | 47.2677841187   | 2837 | -0%    |     (run time too short) 0.7074 |
| TPC-H 9   | 12.3693571091   | 743  | 12.5635232925   | 754  | +2%    |     (run time too short) 0.0000 |
| TPC-H 10  | 9.75266170502   | 586  | 9.68856334686   | 582  | -1%    |     (run time too short) 0.0000 |
| TPC-H 11  | 133.241668701   | 7995 | 134.474884033   | 8069 | +1%    |     (run time too short) 0.0000 |
| TPC-H 12  | 16.3191566467   | 980  | 16.3221416473   | 980  | +0%    |     (run time too short) 0.9078 |
| TPC-H 13  | 16.358877182    | 982  | 16.2909431458   | 978  | -0%    |     (run time too short) 0.0002 |
| TPC-H 14  | 33.1488876343   | 1989 | 33.1110305786   | 1987 | -0%    |     (run time too short) 0.1860 |
| TPC-H 15  | 13.7051477432   | 823  | 13.7839441299   | 828  | +1%    |     (run time too short) 0.0000 |
| TPC-H 16  | 29.5032062531   | 1771 | 29.4258079529   | 1766 | -0%    |     (run time too short) 0.0000 |
| TPC-H 17  | 1.0299077034    | 62   | 1.02835857868   | 62   | -0%    |                          0.0178 |
| TPC-H 18  | 9.68880081177   | 582  | 9.68684101105   | 582  | -0%    |                          0.8580 |
| TPC-H 19  | 15.5609865189   | 934  | 15.4635782242   | 928  | -1%    |     (run time too short) 0.0000 |
| TPC-H 20  | 0.69080722332   | 42   | 0.705316603184  | 43   | +2%    |                          0.0000 |
| TPC-H 21  | 0.0643693879247 | 4    | 0.0638498663902 | 4    | -1%    |        (not enough runs) 0.0000 |
| TPC-H 22  | 72.7827529907   | 4367 | 71.9938278198   | 4320 | -1%    |     (run time too short) 0.0000 |
| average   |                 |      |                 |      | +0%    |                                 |
+-----------+-----------------+------+-----------------+------+--------+---------------------------------+
```

**Comment by mrks** [comment]
> it++ is dangerous.

Could you give an example?

**Comment by mrzzzrm** [comment]
This thing in join_hash_steps:

```c++
        while (it != end) {
          const auto& value = *it;
          ++it;
```

I did this as 


```c++
        while (it != end) {
          const auto& value = *(it++);
```

which created an unnecessary copy of the iterator and took a good chunk out of the JoinHash performance.

File: src/lib/storage/segment_iterate.hpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+#include "storage/segment_iterables/any_segment_iterable.hpp"
+
+/**
+ * This file provides the main entry points to read Segment data, irrespective of the underlying encoding.
+ *
+ * Two main signatures are provided:
+ *      segment_with_iterators[_filtered]()    Calls the functor with a begin and end iterator
+ *      segment_iterate[_filtered]()           Calls the functor with each value in the segment
+ *
+ * Both functions optionally take a PosList which allows for selective access to the values in a segment.
```
Comment (mrks) [inline src/lib/storage/segment_iterate.hpp]: "The ..._filtered versions take a PosList"?

File: src/lib/storage/segment_iterate.hpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * The template parameter T is either (if known to the caller) the DataType of the values contained in the segment, or
```
Comment (mrks) [inline src/lib/storage/segment_iterate.hpp]: Why would the caller pass T?

File: src/lib/operators/jit_operator/operators/jit_read_tuples.cpp (hunk 1 above)
Comment (mrzzzrm) [reply src/lib/operators/jit_operator/operators/jit_read_tuples.cpp]: well, it does add a few loc, but changed it

File: src/lib/operators/sort.cpp (hunk 5 above)
Comment (mrzzzrm) [reply src/lib/operators/sort.cpp]: kk, note that this also means that `SegmentIteratorValue` becomes `SegmentIteratorPosition`

File: src/lib/CMakeLists.txt [hunk 12]
```diff
@@ -458,15 +458,18 @@ set(
...
     storage/segment_encoding_utils.cpp
     storage/segment_encoding_utils.hpp
     storage/segment_iterables.hpp
+    storage/segment_iterables/any_segment_iterable.cpp
     storage/segment_iterables/any_segment_iterable.hpp
     storage/segment_iterables/any_segment_iterator.hpp
     storage/segment_iterables/base_segment_iterators.hpp
     storage/segment_iterables/create_iterable_from_attribute_vector.hpp
     storage/segment_iterables/segment_iterator_values.hpp
+storage/segment_iterate.hpp
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/CMakeLists.txt]: CLion is such a pain right now, I am seriously considering emacs

File: src/lib/operators/sort.cpp (hunk 5 above)
Comment (mrzzzrm) [reply src/lib/operators/sort.cpp]: Decided I like `SegmentPosition` even better.

File: src/lib/storage/reference_segment.cpp [hunk 13]
```diff
@@ -18,13 +18,6 @@ ReferenceSegment::ReferenceSegment(const std::shared_ptr<const Table>& reference
       _pos_list(pos) {
   Assert(_referenced_column_id < _referenced_table->column_count(), "ColumnID out of range")
       DebugAssert(referenced_table->type() == TableType::Data, "Referenced table must be Data Table");
-
-  // Theoretically, a ReferenceSegment can become bigger than the input segments of the operator. This can happen, for
-  // example, if a hash-based join puts all entries into the same bucket and generates a single segment. It is the duty
-  // of the operator to make sure that, once a ReferenceSegment is full, the next one is started. So far, most
-  // operators ignore this, simply because we have not experienced the issue and have not considered it to be a
-  // priority. This assert makes sure that we become aware of it becoming relevant.
-  Assert(pos->size() <= Chunk::MAX_SIZE, "PosList exceeds Chunk::MAX_SIZE");
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/storage/reference_segment.cpp]: No idea what happened there

File: src/lib/storage/segment_iterate.hpp (hunk 11 above)
Comment (mrzzzrm) [reply src/lib/storage/segment_iterate.hpp]: Because he calls the function from code that is already instantiated per DataType

Comment (mrzzzrm) [reply src/lib/storage/segment_iterate.hpp]: See, e.g., _materialize_sort_column

File: src/lib/operators/table_scan/column_vs_column_table_scan_impl.cpp [hunk 14]
```diff
@@ -66,26 +68,21 @@ std::shared_ptr<PosList> ColumnVsColumnTableScanImpl::scan_chunk(ChunkID chunk_i
...
-        auto left_segment_iterable = create_iterable_from_segment<LeftType>(typed_left_segment);
-        auto right_segment_iterable = create_iterable_from_segment<RightType>(typed_right_segment);
-
         // Dirty hack to avoid https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86740
+        const auto left_it_copy = left_it;
+        const auto left_end_copy = left_end;
+        const auto right_it_copy = right_it;
         const auto chunk_id_copy = chunk_id;
-        const auto matches_out_copy = matches_out;
+        const auto& matches_out_copy = matches_out;
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/operators/table_scan/column_vs_column_table_scan_impl.cpp]: "matches_out_ref"

File: src/lib/storage/segment_iterate.hpp (hunk 11 above)
Comment (mrks) [reply src/lib/storage/segment_iterate.hpp]: My fault, I was not clear enough. I would add a comment saying why it would be good for the caller to do this.

File: src/lib/operators/table_scan/column_vs_column_table_scan_impl.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/operators/table_scan/column_vs_column_table_scan_impl.cpp]: yepp

File: src/lib/operators/sort.cpp (hunk 5 above)
Comment (mrks) [reply src/lib/operators/sort.cpp]: Sounds good.

File: src/lib/CMakeLists.txt [hunk 15]
```diff
@@ -466,8 +466,8 @@ set(
     storage/segment_iterables/any_segment_iterator.hpp
     storage/segment_iterables/base_segment_iterators.hpp
     storage/segment_iterables/create_iterable_from_attribute_vector.hpp
-    storage/segment_iterables/segment_iterator_values.hpp
-storage/segment_iterate.hpp
+storage/segment_iterables/segment_positions.hpp
```
Comment (mrks) [inline src/lib/CMakeLists.txt]: ...

Comment (mrzzzrm) [reply src/lib/CMakeLists.txt]: do your lint magic, I suppose.
